[Enhancement] Optimize shard lock acquisition to reduce the holding time

srlch opened the pull request:
## Why I'm doing:
In current impl, tablet meta lock or some primary key relative locks will be try to acquired after the shard lock has been holded in some place.
This may cause the shard lock to be held for a long time and cannot be released.

## What I'm doing:
Try to release the shard lock before try to acquire another lock if it is possible.

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.4
  - [x] 3.3
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0

kevincai commented on be/src/storage/tablet.cpp:
```diff
@@ -1534,6 +1534,13 @@ bool Tablet::_contains_rowset(const RowsetId rowset_id) {
+    // in TabletUpdates which means that those lock context can be sperated from
+    // the tablet meta_lock when trying to get the extra info
+    if (_updates) {
+        _updates->get_tablet_info_extra(tablet_info);
```
if this is the case, shall we move this to the end and execute without meta lock? the seq will be the same  as the original one, just without the meta lock.

kevincai commented on be/src/storage/tablet_manager.cpp:
```diff
@@ -983,28 +983,35 @@ Status TabletManager::report_all_tablets_info(std::map<TTabletId, TTablet>* tabl
     StarRocksMetrics::instance()->report_all_tablets_requests_total.increment(1);
 
-    size_t max_tablet_rowset_num = 0;
+    std::vector<TabletSharedPtr> all_tablets;
```
may cause this vector size too large. move this to tablet_shard level. so the under the lock, just get the array from the tablet_shard, and then generate the report info without lock.

This way the vector size will be the max(tablets_shards) instead of sum(tablets_shards);

srlch replied:
updated

srlch committed:
[Enhancement] Optimize shard lock acquisition to reduce the holding time

## Why I'm doing:
In current impl, tablet meta lock or some primary key relative locks will be try to acquired after the shard lock has been holded in some place.
This may cause the shard lock to be held for a long time and cannot be released.

## What I'm doing:
Try to release the shard lock before try to acquire another lock if it is possible.

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

yongbingwang commented:
@Mergifyio backport branch-3.2.15-0219
